// @aws-cdk/aws-robomaker/lib/robomaker-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoboMakerMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class RoboMakerMetrics {
    static vCpuAverage(dimensions) {
        return {
            namespace: 'AWS/RoboMaker',
            metricName: 'vCPU',
            dimensions,
            statistic: 'Average',
        };
    }
    static memoryMaximum(dimensions) {
        return {
            namespace: 'AWS/RoboMaker',
            metricName: 'Memory',
            dimensions,
            statistic: 'Maximum',
        };
    }
    static realTimeRatioAverage(dimensions) {
        return {
            namespace: 'AWS/RoboMaker',
            metricName: 'RealTimeRatio',
            dimensions,
            statistic: 'Average',
        };
    }
    static simulationUnitSum(dimensions) {
        return {
            namespace: 'AWS/RoboMaker',
            metricName: 'SimulationUnit',
            dimensions,
            statistic: 'Sum',
        };
    }
}
exports.RoboMakerMetrics = RoboMakerMetrics;
