// @aws-cdk/aws-appmesh/lib/client-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
var CertificateType;
(function (CertificateType) {
    CertificateType["ACMPCA"] = "acm";
    CertificateType["FILE"] = "file";
})(CertificateType || (CertificateType = {}));
/**
 * (experimental) Defines the TLS validation context trust.
 *
 * @experimental
 */
class ClientPolicy {
    /**
     * (experimental) Tells envoy where to fetch the validation context from.
     *
     * @experimental
     */
    static fileTrust(props) {
        return new ClientPolicyImpl(props.ports, CertificateType.FILE, props.certificateChain, undefined);
    }
    /**
     * (experimental) TLS validation context trust for ACM Private Certificate Authority (CA).
     *
     * @experimental
     */
    static acmTrust(props) {
        return new ClientPolicyImpl(props.ports, CertificateType.ACMPCA, undefined, props.certificateAuthorities);
    }
}
exports.ClientPolicy = ClientPolicy;
_a = JSII_RTTI_SYMBOL_1;
ClientPolicy[_a] = { fqn: "@aws-cdk/aws-appmesh.ClientPolicy", version: "1.94.1" };
class ClientPolicyImpl extends ClientPolicy {
    constructor(ports, certificateType, certificateChain, certificateAuthorityArns) {
        super();
        this.ports = ports;
        this.certificateType = certificateType;
        this.certificateChain = certificateChain;
        this.certificateAuthorityArns = certificateAuthorityArns;
    }
    bind(_scope) {
        var _b, _c;
        if (this.certificateType === CertificateType.ACMPCA && ((_b = this.certificateAuthorityArns) === null || _b === void 0 ? void 0 : _b.map(certificateArn => certificateArn.certificateAuthorityArn).length) === 0) {
            throw new Error('You must provide at least one Certificate Authority when creating an ACM Trust ClientPolicy');
        }
        else {
            return {
                clientPolicy: {
                    tls: {
                        ports: this.ports,
                        validation: {
                            trust: {
                                [this.certificateType]: this.certificateType === CertificateType.FILE
                                    ? {
                                        certificateChain: this.certificateChain,
                                    }
                                    : {
                                        certificateAuthorityArns: (_c = this.certificateAuthorityArns) === null || _c === void 0 ? void 0 : _c.map(certificateArn => certificateArn.certificateAuthorityArn),
                                    },
                            },
                        },
                    },
                },
            };
        }
    }
}
