// @aws-cdk/aws-synthetics/lib/synthetics-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudWatchSyntheticsMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class CloudWatchSyntheticsMetrics {
    static _2XxSum(dimensions) {
        return {
            namespace: 'CloudWatchSynthetics',
            metricName: '2xx',
            dimensions,
            statistic: 'Sum',
        };
    }
    static _4XxSum(dimensions) {
        return {
            namespace: 'CloudWatchSynthetics',
            metricName: '4xx',
            dimensions,
            statistic: 'Sum',
        };
    }
    static _5XxSum(dimensions) {
        return {
            namespace: 'CloudWatchSynthetics',
            metricName: '5xx',
            dimensions,
            statistic: 'Sum',
        };
    }
    static durationAverage(dimensions) {
        return {
            namespace: 'CloudWatchSynthetics',
            metricName: 'Duration',
            dimensions,
            statistic: 'Average',
        };
    }
    static failedSum(dimensions) {
        return {
            namespace: 'CloudWatchSynthetics',
            metricName: 'Failed',
            dimensions,
            statistic: 'Sum',
        };
    }
    static failedRequestsSum(dimensions) {
        return {
            namespace: 'CloudWatchSynthetics',
            metricName: 'Failed requests',
            dimensions,
            statistic: 'Sum',
        };
    }
    static successPercentAverage(dimensions) {
        return {
            namespace: 'CloudWatchSynthetics',
            metricName: 'SuccessPercent',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.CloudWatchSyntheticsMetrics = CloudWatchSyntheticsMetrics;
