// @aws-cdk/aws-mediapackage/lib/mediapackage-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaPackageMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class MediaPackageMetrics {
    static egressRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/MediaPackage',
            metricName: 'EgressRequestCount',
            dimensions,
            statistic: 'Sum',
        };
    }
    static egressResponseTimeAverage(dimensions) {
        return {
            namespace: 'AWS/MediaPackage',
            metricName: 'EgressResponseTime',
            dimensions,
            statistic: 'Average',
        };
    }
    static egressBytesSum(dimensions) {
        return {
            namespace: 'AWS/MediaPackage',
            metricName: 'EgressBytes',
            dimensions,
            statistic: 'Sum',
        };
    }
}
exports.MediaPackageMetrics = MediaPackageMetrics;
