// @aws-cdk/aws-backup/lib/resource.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupResource = exports.TagOperation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("@aws-cdk/core");
/**
 * (experimental) An operation that is applied to a key-value pair.
 *
 * @experimental
 */
var TagOperation;
(function (TagOperation) {
    TagOperation["STRING_EQUALS"] = "STRINGEQUALS";
    TagOperation["DUMMY"] = "dummy";
})(TagOperation = exports.TagOperation || (exports.TagOperation = {}));
/**
 * (experimental) A resource to backup.
 *
 * @experimental
 */
class BackupResource {
    /**
     * @experimental
     */
    constructor(resource, tagCondition, construct) {
        this.resource = resource;
        this.tagCondition = tagCondition;
        this.construct = construct;
    }
    /**
     * (experimental) Adds all supported resources in a construct.
     *
     * @param construct The construct containing resources to backup.
     * @experimental
     */
    static fromConstruct(construct) {
        return new BackupResource(undefined, undefined, construct);
    }
    /**
     * (experimental) A DynamoDB table.
     *
     * @experimental
     */
    static fromDynamoDbTable(table) {
        return BackupResource.fromArn(table.tableArn);
    }
    /**
     * (experimental) An EC2 instance.
     *
     * @experimental
     */
    static fromEc2Instance(instance) {
        return BackupResource.fromArn(core_1.Stack.of(instance).formatArn({
            service: 'ec2',
            resource: 'instance',
            resourceName: instance.instanceId,
        }));
    }
    /**
     * (experimental) An EFS file system.
     *
     * @experimental
     */
    static fromEfsFileSystem(fileSystem) {
        return BackupResource.fromArn(core_1.Stack.of(fileSystem).formatArn({
            service: 'elasticfilesystem',
            resource: 'file-system',
            resourceName: fileSystem.fileSystemId,
        }));
    }
    /**
     * (experimental) A RDS database instance.
     *
     * @experimental
     */
    static fromRdsDatabaseInstance(instance) {
        return BackupResource.fromArn(instance.instanceArn);
    }
    /**
     * (experimental) A list of ARNs or match patterns such as `arn:aws:ec2:us-east-1:123456789012:volume/*`.
     *
     * @experimental
     */
    static fromArn(arn) {
        return new BackupResource(arn);
    }
    /**
     * (experimental) A tag condition.
     *
     * @experimental
     */
    static fromTag(key, value, operation) {
        return new BackupResource(undefined, {
            key,
            value,
            operation,
        });
    }
}
exports.BackupResource = BackupResource;
_a = JSII_RTTI_SYMBOL_1;
BackupResource[_a] = { fqn: "@aws-cdk/aws-backup.BackupResource", version: "1.94.1" };
