// @aws-cdk/aws-inspector/lib/inspector-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.InspectorMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class InspectorMetrics {
    static totalHealthyAgentsAverage(dimensions) {
        return {
            namespace: 'AWS/Inspector',
            metricName: 'TotalHealthyAgents',
            dimensions,
            statistic: 'Average',
        };
    }
    static totalAssessmentRunsAverage(dimensions) {
        return {
            namespace: 'AWS/Inspector',
            metricName: 'TotalAssessmentRuns',
            dimensions,
            statistic: 'Average',
        };
    }
    static totalMatchingAgentsAverage(dimensions) {
        return {
            namespace: 'AWS/Inspector',
            metricName: 'TotalMatchingAgents',
            dimensions,
            statistic: 'Average',
        };
    }
    static totalFindingsAverage(dimensions) {
        return {
            namespace: 'AWS/Inspector',
            metricName: 'TotalFindings',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.InspectorMetrics = InspectorMetrics;
